python/ray/job_config.py

```diff
@@ -1,81 +1,131 @@
+            can be "detached" or "non_detached".
     """
 
     def __init__(
         self,
-        jvm_options: List[str] = None,
-        code_search_path: List[str] = None,
-        runtime_env: dict = None,
-        client_job: bool = False,
+        jvm_options: Optional[List[str]] = None,
+        code_search_path: Optional[List[str]] = None,
+        runtime_env: Optional[dict] = None,
         metadata: Optional[dict] = None,
         ray_namespace: Optional[str] = None,
         default_actor_lifetime: str = "non_detached",
-        py_driver_sys_path: List[str] = None,
+        _client_job: bool = False,
+        _py_driver_sys_path: List[str] = None,
     ):
+        #: The jvm options for java workers of the job.
         self.jvm_options = jvm_options or []
+        #: A list of directories or jar files that
+        #: specify the search path for user code.
         self.code_search_path = code_search_path or []
         # It's difficult to find the error that caused by the
         # code_search_path is a string. So we assert here.
         assert isinstance(self.code_search_path, (list, tuple)), (
             f"The type of code search path is incorrect: " f"{type(code_search_path)}"
         )
-        self.client_job = client_job
+        self._client_job = _client_job
+        #: An opaque metadata dictionary.
         self.metadata = metadata or {}
+        #: A namespace is a logical grouping of jobs and named actors.
         self.ray_namespace = ray_namespace
         self.set_runtime_env(runtime_env)
         self.set_default_actor_lifetime(default_actor_lifetime)
-        self.py_driver_sys_path = py_driver_sys_path or []
+        # A list of directories that specify the search path for python workers.
+        self._py_driver_sys_path = []
 
     def set_metadata(self, key: str, value: str) -> None:
+        """Add key-value pair to the metadata dictionary.
+
+        If the key already exists, the value will be overwritten to the new value.
+
+        Examples:
+            .. testcode::
+
+                import ray
+                from ray.job_config import JobConfig
+
+                job_config = JobConfig()
+                job_config.set_metadata("submitter", "foo")
+
+        Args:
+            key: The key of the metadata.
+            value: The value of the metadata.
+        """
         self.metadata[key] = value
 
-    def serialize(self):
+    def _serialize(self) -> str:
         """Serialize the struct into protobuf string"""
-        return self.get_proto_job_config().SerializeToString()
+        return self._get_proto_job_config().SerializeToString()
 
     def set_runtime_env(
         self,
-        runtime_env: Optional[Union[Dict[str, Any], "RuntimeEnv"]],  # noqa: F821
+        runtime_env: Optional[Union[Dict[str, Any], "RuntimeEnv"]],
         validate: bool = False,
     ) -> None:
         """Modify the runtime_env of the JobConfig.
```

Comment:
Is it useful to describe when or why a user would want to modify the runtime_env of the JobConfig?

Reply:
I think it's just an alternative way to set runtime_env other than via the constructor. But there is no particular reason why people have to use this way.